external/fv3fit/fv3fit/keras/_models/shared/pure_keras.py

```diff
@@ -175,3 +176,80 @@ def input_sensitivity(self, stacked_sample: xr.Dataset) -> InputSensitivity:
         std_factors = {name: np.std(data, axis=0) for name, data in data_dict.items()}
         jacobians_std = nondimensionalize_jacobians(jacobians, std_factors)
         return InputSensitivity(jacobians=jacobians_std)
+
+
+@io.register("all-keras-dict")
+class PureKerasDictPredictor(Predictor):
+    """Save a keras model encoding the input/output names into a saved_model
+    artifact
+
+    The output names of the keras model are assumed to be correct. To do this
+    requires including the following logic when building the model::
+
+        out_with_correct_name = tf.keras.layers.Lambda(
+            lambda x: x, name="desired_name")
+
+    Note:
+        This class relies on PureKerasModel to implement Mhe Predictor methods.
+        It might be worth refactoring the implementation here instead since the
+        dict-in/out format has more structure.
+
+    """
+
+    _MODEL_FILENAME = "model.tf"
+    _CONFIG_FILENAME = "config.yaml"
+
+    def __init__(self, model: tf.keras.Model, passthrough: Any = None):
+        """Initialize the predictor.
+
+        Args:
+            model: a keras model. Tuple in/out models are okay, but will be cast
+                to a dict-output model before saving.
+            passthru: any data! Is NOT serialized, but can be
+                used by other parts of the training workflow. Mostly exists for
+                backwards compatibility and should be avoided in new code.
+        """
+        dict_model = ensure_dict_output(model)
+        tuple_model = ensure_tuple_model(dict_model)
+        super().__init__(tuple_model.input_names, tuple_model.output_names)
+        self._model = tuple_model
+        self._pure_keras_predictor = PureKerasModel(
+            self.input_variables,
+            self.output_variables,
+            tuple_model,
+            # unstacked dims is hardcoded most places in fv3fit.keras, but it
+            # seems like useful metadata, but more of an aspect of the training
+            # data pipeline since it is the piece of code that removes the
+            # dimension names...would be nice to remove this concept...e.g. pass
+            # unstacked_dims to the .dump rather than in the constructor
+            # TODO consider this refactor
```

Comment:
`unstacked_dims` needs to be passed alongside the tf.keras.Model instance, since if you have a xr.Dataset and a tf.keras.Model you need the additional metadata of what are the dimensions the model is expecting (it's `["sample"] + unstacked_dims` where sample is the stacking of anything not in unstacked_dims). A PureKerasModel initialized without `unstacked_dims` cannot `predict`.

Reply:
I understand. I just chose to put the "z" hardcode here rather than in the training func as you did here: https://github.com/ai2cm/fv3net/blob/04c4d4c358f6f00f635c6803da14046200084ec8/external/fv3fit/fv3fit/keras/_models/dense.py#L230

I think unstacked dims should be passed to dump. The hardcodes exist because the source data has been stripped of this metadata by the time the model is built. The thing that dumps the model knows what the dimensions are though, so I think this would work nicely:
```
predictor.dump(path, unstacked_dims)
```

Reply:
Or maybe we'd have to make a PredictorFactory...ugh:
```
predictor_factory = PredictorFactory(model, inputs, outputs)
# ...in the higher scope
predictor = predictor_factor.build(unstacked_dims)
predictor.dump(path)
```

Reply:
would be easy to do with `functools.partial` or curry.

Reply:
Does the thing that dumps the code know the dimensions? I thought `.dump` gets called in `fv3fit.train`, but the stacked dimensions are only known inside the context of a training function (where they are assumed). We'd have to extract the unstacked dimensions from the training function into the top level to pass this directly to `dump`.

Specializing the predictor on column prediction by passing "z" explicitly here is fine. We can always refactor it if we use this for more than columnwise prediction (e.g. the convolutional model).

Reply:
We should talk more if we want to later implement a change like this - as I understand it, the scope in `fv3fit.train` where `dump` is called does not have access to / any dependence on `unstacked_dims`.

Reply:
I usually find the pulling info that needs to be shared to the mini-max scope and injecting it where needed is a good option. Lot's of ways to do it...maybe the train funcs get passed it as an arg if they should know about the dimension metadata.

Reply:
It feels like hardcoding it in the training function is fine because this information is already contextually hardcoded everywhere. e.g. the dense training function assumes its arrays are [sample, z] when it does clipping or when it does model building, and the convolutional training assumes the dimension order when it applies halo updates. In the training function, there isn't a need to make the dimensions arbitrary and it keeps things much simpler if they're not. Kind of like when you decorate the required dims for xarray apply_ufuncs.

Reply:
Another way to say this is, the information flow would be in the opposite direction. The training functions are all hard-coded to a specific unstacked_dims, and would need to pass this upwards to `fv3fit.train` if it needed it (e.g. to assert the data being loaded has the correct unstacked_dims).

Reply:
That does make sense. Thanks for the explanation.